### PR TITLE
Ruleset Model Description Phase 2

### PR DESCRIPTION
**Pull request by jcyuan2020**
Pull request overview
---------------------
 - Implements new tables and/or adds new tabular reporting entries for Std 229 Ruleset models;
 - Original NFP: #10051
 - Also, if the original NFP should be included in the final code, then PR #10051 should be merged in (to `develop` branch) first prior to merging in the current PR.

Diffs: 
- Massive tabular diffs are expected since the tabular output reports are now changed in many places;

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by JasonGlazer** [comment]
@jcyuan2020 is the Win64 build failure a real problem with the code or is it just a glitch in the CI?

**Comment by jcyuan2020** [comment]
> @jcyuan2020 is the Win64 build failure a real problem with the code or is it just a glitch in the CI?

From the build error message and the mass failure messages this should belong to some kind of a glitch in the CI. Anyway I just pushed some new changes and this will trigger a new round of CI build to see if the problem persists.

File: src/EnergyPlus/BoilerSteam.cc
Comment (rraustad) [inline src/EnergyPlus/BoilerSteam.cc:549]: Wouldn't it be better to move these 2 big blocks of code, the one above and the one below, to below this big IF/ELSE block as 1 block of code protected with `if (state.dataPlnt->PlantFinalSizesOkayToReport) {`

Comment (jcyuan2020) [reply src/EnergyPlus/BoilerSteam.cc:549]: Thanks! That's seems to be a valid thought to reduce the code repetition. One little issue here is that the conditions for these two block do not seem to be perfectly aligned---the second one has an additional upper level condition to get into the blocks of `if (state.dataPlnt->PlantFinalSizesOkayToReport) {` as follows:
```
                    if (this->NomCap > 0.0 && tmpNomCap > 0.0) {
                        Real64 NomCapUser = this->NomCap;
                        if (state.dataPlnt->PlantFinalSizesOkayToReport) {
```

It is possible to combine these two added std 229 blocks into one, (literally) by adding a common condition to replicate the logic using `if(state.dataPlnt->PlantFinalSizesOkayToReport && (this->NomCapWasAutoSized || (!this->NomCapWasAutoSized && (this->NomCap > 0.0 && tmpNomCap > 0.0)))) {`. But it seems to somehow ruin the readability of the logic when doing this way. 

For this reason I would prefer the original version to keep the logic more readable (and also more closely attached to the original logic flow of doing the two `BaseSizer::reportSizerOutput()` branches.

**Comment by Myoldmopar** [comment]
There are some build warnings showing in this branch. They are minor, but will need to get fixed before this gets merged. https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-bfc7860d091cef1ece9e50ad931d88afc58fcc3e-PR10143-x86_64-Linux-Ubuntu-22.04-gcc-11.4.html

Also, as @mjwitte mentioned yesterday, because this branch changes the "structure" of the output tables, it will need some notes in the `/src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md` file.  I'll give this a look in the meantime, but this won't be merged until then.

**Comment by rraustad** [comment]
@jcyuan2020 can you please paste one of the new tables here to document what they look like with this branch.

**Comment by jcyuan2020** [comment]
> @jcyuan2020 can you please paste one of the new tables here to document what they look like with this branch.

Here is an example table for chillers:

image

**Comment by rraustad** [comment]
Hmm, why is the COP chiller Rated and Reference conditions = 0? and the other chiller type shows values.

image

**Comment by jcyuan2020** [comment]
> Hmm, why is the COP chiller Rated and Reference conditions = 0? and the other chiller type shows values.
> 
> image

That's strange---I checked the variables links (`TempDesCondIn` and `TempDesEvapOut`) in the code and they are correct. 

The case pasted above is from "CoolingTower" and the ConstantCOP chiller does have a condenser loop. I checked the constant COP chiller code and it seems that this is due to the fact that the `Chiller:ConstantCOP` object does not have a design condenser in nor a design evap out input field--thus the zero initialization of these variables are never touched since creation. A possible solution would be adding a default 29.44C and 6.67 for this as default values. But this should be beyond the scope for the current PR and I will post another issue and PR to fix that.

**Comment by Myoldmopar** [comment]
Of course Decent CI machine had to reboot now :)  I restarted it but it might be a while before results pop up.  I'll take a look locally and try it out.  It would be nice to get this merged in soon.

**Comment by Myoldmopar** [comment]
Build warnings appear gone locally.  What's the deal with the big ESO diffs?  Is it expected here?  Or just behind develop?  I'll run regressions after the unit/integration test suite finishes.

**Comment by Myoldmopar** [comment]
Alright, regressions look better locally; I think they were showing up because of being out of date with develop.  Locally I see lots of audit diffs, and EIO diffs that are new reports coming out.

File: src/EnergyPlus/HeatRecovery.cc
Comment (Myoldmopar) [inline src/EnergyPlus/HeatRecovery.cc:1703]: This feels a bit repeated in all these places, but I don't think there's time to sweat that right now.

File: src/EnergyPlus/HeatingCoils.cc
Comment (Myoldmopar) [inline src/EnergyPlus/HeatingCoils.cc:1864]: Why is this in here but commented?  I may need to read back through the PR to understand why.

File: src/EnergyPlus/MixedAir.cc
Comment (Myoldmopar) [inline src/EnergyPlus/MixedAir.cc]: Another addition but commented?

**Review by Myoldmopar — APPROVED** [review]
I think this is good for now.  I understand there is still a bit of work to go on rulesets that will continue beyond this PR, so I'm not going to scrutinize this too much now, especially given the release time.

**Comment by Myoldmopar** [comment]
@rraustad you OK with this going in as-is?  I think we can deal with the zero table values in a later PR and get this in for the freeze.

**Comment by jcyuan2020** [comment]
@Myoldmopar The diffs should be largely expected since the outputs are changed pretty much for every test file. Also, I can do a final clean up if you can wait (maybe a couple of hours?).

**Comment by Myoldmopar** [comment]
> The diffs should be largely expected

I would not expect numeric outputs to change and cause math diffs.  But file diffs due to added outputs, yes, those are expected and fine.  And that's what I see locally.



> I can do a final clean up

What do you need to clean up?

**Comment by rraustad** [comment]
>@rraustad you OK with this going in as-is? I think we can deal with the zero table values in a later PR and get this in for the freeze.

Yes, I'm OK with merging this. I looked over it several times.

**Comment by jcyuan2020** [comment]
> 

@Myoldmopar For the clean up, my plan was to remove some variables that was not implemented for the current stage (so the empty columns); also cleaning up some commented-out lines if needed.

**Comment by Myoldmopar** [comment]
OK, @jcyuan2020 go ahead and do your cleanup pass.  I'm going to move onto other PRs to review then.

**Comment by jcyuan2020** [comment]
> OK, @jcyuan2020 go ahead and do your cleanup pass. I'm going to move onto other PRs to review then.

@Myoldmopar Thanks! Working on it now.

**Comment by jcyuan2020** [comment]
@Myoldmopar Just let you know that the clean-up was done and changes were pushed in a few hours ago.

**Comment by mjwitte** [comment]
UnitarySystem_MultiSpeedDX_EconoStaging.idf is showing a big table diff, but I can't reproduce it locally on windows. I suspect there's an uninitialized variable feeding into the heat rejection total that only shows up on linux? Mac diffs are clean.
image

File: src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md
Comment (mjwitte) [inline src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md:19]: This background isn't necessary, but it's fine here.

File: src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md
Comment (mjwitte) [inline src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md]: It appears that this added columns to many of the subtables in the EquipmentSummary report, not just Fans. I realize listing every table and column name may be overkill, but this should at least mention the tables/subtables that have changed and which ones are new.

While editing this file, please fix the typos in the EIO System Sizing section:
Missing unit is added to the EIO ystem Sizing Informationg ta

File: src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md
Comment (mjwitte) [inline src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md:30]: This is not a new table.

File: src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md
Comment (mjwitte) [inline src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md]: I don't think this was completed.

**Review by mjwitte — COMMENTED** [review]
I ran UnitarySystem_MultiSpeedDX_EconoStaging to see the bit table diffs (which don't show up locally on windows) but then I was comparing this branch vs develop for other reports like the equipment summary. Some of the new columns need fixing. e.g. 
image

Unknown should be N/A?

image

Column headers should all be Title Case (needs fixing in other tables, too).
Purpose is wrong.
Is Autosized should be Yes/No
Airloop Name should be known here.

image

File: src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md
Comment (jcyuan2020) [reply src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md]: The file is now updated.

**Comment by jcyuan2020** [comment]
> I ran UnitarySystem_MultiSpeedDX_EconoStaging to see the bit table diffs (which don't show up locally on windows) but then I was comparing this branch vs develop for other reports like the equipment summary. Some of the new columns need fixing. e.g. image
> 
> Unknown should be N/A?

The "unknown" was the default assignment in the original code in the initialization, so that would show up for a case where no plantLoopName is fed. Now I changed to use the original way where the plantLoopName was fed in the code and it should be "N/A" when the loop is not applicable.
 
> Column headers should all be Title Case (needs fixing in other tables, too). 

Column headers are now revised to show Title Case.

> Purpose is wrong. 

I still could not find a proper field in the current struct (and the current location) that could represent the "Purpose" or "placement"? Current locations are [here](https://github.com/NREL/EnergyPlus/pull/10143/files#diff-737f152cfc96fe6b347b1793b736ad42fc93d2f282532c533c6212d073edcda7R1451-R1455), and [here](https://github.com/NREL/EnergyPlus/pull/10143/files#diff-14a49211ba32816f868594de296cc9e60d17b1eb9dc9377d04d54ec34a888edfR317-R320).  

> Is Autosized should be Yes/No 

Now it is "Yes" and "No".

> Airloop Name should be known here.

That's odd---it shows "N/A" because the Fan's `AirLoopNum` is 0.

File: src/EnergyPlus/OutputReportPredefined.cc
Comment (mjwitte) [inline src/EnergyPlus/OutputReportPredefined.cc:842]: This ended up as a subtable to the Coil Sizing Details table. Per the output rules, it should be in Equipment Summary? `s->pdrEquip` @JasonGlazer would you agree?

File: src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md
Comment (mjwitte) [inline src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md]: Equipment Summary - Coil Connections

File: src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md
Comment (mjwitte) [inline src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md:23]: Is it accurate to say that all new columns have been added to the right of existing columns? That's important to know for interfaces or postprocessors which read the tables by row and column.

**Review by mjwitte — COMMENTED** [review]
@jcyuan2020 Thanks for the cleanups.

For the fan "Purpose" column it should be blank for now and then we can populate it after I/O Freeze. I recall that you were using the object name as a placeholder during development. If there are any other columns like that, we need to determine if we can populate them later (so make them blank for now) or if they column should be removed.

@rraustad Fans should know their placement for sizing purposes, yes? (Supply, Return, Exhaust).

**Comment by mjwitte** [comment]
@jcyuan2020 This table still needs some help, heading and content.
image

**Comment by mjwitte** [comment]
@jcyuan2020 Ran HospitalLowEnergy to see more equipment types. 
- Pumps "Is Autosized" is stil 1 or 0.
- For PlantLoop or CondenserLoop and Air Terminals, the "Name" column is redundant.
- For PlantLoop or CondenserLoop the column heading "Type (PlantLoop or CondenserLoop)" should just be "Type".
- Same table "Provides Hooling" --> Provides Cooling
- And maybe the Fan "Purpose" column should just be one word. I don't think there are other places in the table output that list the possibilities in the column heading.
- Running testfile FluidCoolerTwoSpeed, it has two plant loops, but the PlantLoop table is empty? AirTerminals is empty also.

@JasonGlazer The order of columns is not consistent across equipment types. For some equipment (e.g. cooling towers), the loop and branch name are first, for others it's in the middle (chillers), and then Pumps have it at the end. I'm sure the RMD script won't care, but for human readers, do we want to have some common approach for ordering the types of columns?

**Comment by rraustad** [comment]
> @rraustad Fans should know their placement for sizing purposes, yes? (Supply, Return, Exhaust).

The fan doesn't really care where it is. Either the parent, or the air loop, or the OA system will set CurZoneEqNum, CurSysNum, or CurOASysNum and then the fan sizes accordingly (usually in SystemAirFlowSizing.cc).

The fan type and index is saved in the PrimaryAirSystem struc as a supFan or retFan. I don't think we save this in the OA system yet, or look at exhaust fans since they don't affect component sizing. The parents also know fan type and name.

CoolingCapacitySizing.cc

    switch (this->primaryAirSystem(this->curSysNum).supFanModelType) {
          case DataAirSystems::StructArrayLegacyFanModels: {
       ... snip...
    switch (this->primaryAirSystem(this->curSysNum).retFanModelType) {
          case DataAirSystems::StructArrayLegacyFanModels: {

The blow or draw thru configuration is also known:

    if (this->primaryAirSystem(this->curSysNum).supFanLocation == DataAirSystems::FanPlacement::BlowThru) {


    struct DefinePrimaryAirSystem // There is an array of these for each primary air system
    {

        FanModelType supFanModelType = FanModelType::Invalid; // indicates which type of fan model to call for supply fan, legacy or new OO
        int SupFanNum = 0;       // index of the supply fan in the Fan data structure when model type is StructArrayLegacyFanModels
        int supFanVecIndex = -1; // index in fan object vector for supply fan when model type is ObjectVectorOOFanSystemModel, zero-based index
        FanPlacement supFanLocation = FanPlacement::Invalid;  // location of fan relative to coil
        FanModelType retFanModelType = FanModelType::Invalid; // indicates which type of fan model to call for return fan, legacy or new OO
        int RetFanNum = 0;           // index of the return fan in the Fan data structure when model type is StructArrayLegacyFanModels
        int retFanVecIndex = -1;     // index in fan object vector for return fan when model type is ObjectVectorOOFanSystemModel, zero-based index
        Real64 FanDesCoolLoad = 0.0; // design fan heat gain for the air loop [W]

**Comment by JasonGlazer** [comment]
> The order of columns is not consistent across equipment types. For some equipment (e.g. cooling towers), the loop and branch name are first, for others it's in the middle (chillers), and then Pumps have it at the end. I'm sure the RMD script won't care, but for human readers, do we want to have some common approach for ordering the types of columns?

@mjwitte @jcyuan2020 it would be good if the order of the columns was consistent.

**Comment by mjwitte** [comment]
@jcyuan2020 Getting better. The air terminal setpoints are now "0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.0000......". If we don't have a good handle on how to populate this, perhaps it should be removed?

At this point, let's focus on getting the column heading order finished, depopulating anything that's showing garbage and leave other fixes for after I/O Freeze.

**Comment by jcyuan2020** [comment]
> @jcyuan2020 Getting better. The air terminal setpoints are now "0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.0000......". If we don't have a good handle on how to populate this, perhaps it should be removed?

I am leaving it as it is with a 0.00 (2 decimal digit) formatting. The variable arrays should be the correct ones. Not sure why in this test file it shows all zero numbers.

> At this point, let's focus on getting the column heading order finished, depopulating anything that's showing garbage and leave other fixes for after I/O Freeze.

For this one, do we care to change column orders for existing elements in existing tables (e.g. Pumps tables)? If we want to preserve the original order of the existing variables in existing tables, the option would be to place the plant loop information to the end; the other two tables (chillers and cooling/fluid coolers) are new tables so we can rearrange with more flexibility.

Otherwise, we can place the plant loop information to near the front of the table. This will involve some changes on existing variables' orders in existing tables (e.g. pumps). And also it would null the statement in the OutputChanges_xxx.md file about that new variables are added to the right of the existing columns (tables).

**Comment by mjwitte** [comment]
> > @jcyuan2020 Getting better. The air terminal setpoints are now "0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.000000; 0.0000......". If we don't have a good handle on how to populate this, perhaps it should be removed?
> 
> I am leaving it as it is with a 0.00 (2 decimal digit) formatting. The variable arrays should be the correct ones. Not sure why in this test file it shows all zero numbers.

I don't understand, why is there a long list of values? @JasonGlazer is this required? Could we just show max/min?

> 
> > At this point, let's focus on getting the column heading order finished, depopulating anything that's showing garbage and leave other fixes for after I/O Freeze.
> 
> For this one, do we care to change column orders for existing elements in existing tables (e.g. Pumps tables)? If we want to preserve the original order of the existing variables in existing tables, the option would be to place the plant loop information to the end; the other two tables (chillers and cooling/fluid coolers) are new tables so we can rearrange with more flexibility.
> 
I think the location info should go to the far right. It's not super-useful to someone browsing the table, IMHO.  And all existing tables should preserve the order and add new info to the right.

**Comment by jcyuan2020** [comment]
> > 
> > I am leaving it as it is with a 0.00 (2 decimal digit) formatting. The variable arrays should be the correct ones. Not sure why in this test file it shows all zero numbers.
> 
> I don't understand, why is there a long list of values? @JasonGlazer is this required? Could we just show max/min?

I think the array is the dayli setpoint series for the supply? But anyway, I found another two variables that seem better to represent the zone supply cooling and heating setpoints; and they does cleared the table nicely (and values look reasonable as well).
 
> > > At this point, let's focus on getting the column heading order finished, depopulating anything that's showing garbage and leave other fixes for after I/O Freeze.
> > 
> > 
> > For this one, do we care to change column orders for existing elements in existing tables (e.g. Pumps tables)? If we want to preserve the original order of the existing variables in existing tables, the option would be to place the plant loop information to the end; the other two tables (chillers and cooling/fluid coolers) are new tables so we can rearrange with more flexibility.
> 
> I think the location info should go to the far right. It's not super-useful to someone browsing the table, IMHO. And all existing tables should preserve the order and add new info to the right.

This is done now.

Also, units are added to column headers whenever applicable.

File: src/EnergyPlus/OutputReportPredefined.cc
Comment (mjwitte) [inline src/EnergyPlus/OutputReportPredefined.cc]: Upper-case [C]

**Review by mjwitte — APPROVED** [review]
@jcyuan2020 Thanks for all the cleanups. There's one lower-case [c] that slipped in. Otherwise, I'm ok with this as-is.
@Myoldmopar I'll let you do a final pass and merge.

**Comment by Myoldmopar** [comment]
One file showing table diffs which is a known item.  Otherwise this has been scrutinized sufficiently.  I'll run quick tests with develop pulled in and if good then merge.

**Comment by Myoldmopar** [comment]
I'm pulling develop to get results with the steam merge, and if happy, this will merge next.

**Comment by Myoldmopar** [comment]
Built and tested locally with latest develop, all good.  Regressions locally confirm what Decent saw last time it ran, which is minor audit and EIO diffs due to new sizing outputs.  I feel very good about this.  I'm pushing up my conflict resolved commit and will let GHA start up, then probably merge shortly.

**Comment by Myoldmopar** [comment]
GH seems kinda stuck waiting on this.  I'm going to merge it in.  Thanks @jcyuan2020 for this work.  Looking forward to how this improves the whole system of rulesets around EnergyPlus.

**Comment by JasonGlazer** [comment]
Many thanks to @jcyuan2020, @Myoldmopar, @mjwitte and @rraustad!

**Pull request merged**
